### PR TITLE
Improve metadata handling

### PR DESCRIPTION
**Pull request by jason-fox**
- [x] Add NGSI-LD Actuation `datasetId` and `metadata` support for Registrations
- [x] Demonstrate that NGSI-v2 already supports `metadata` for Registrations
- [x] Add metadata handling to NGSI-v2 notifications
- [x] Add NGSI-LD Actuation `metadata` support for Notifications
- [x]  Add NGSI-LD Actuation `datasetId` for Notifications

### CODE REVIEW

**Comment by jason-fox** [comment]
@mapedraza @fgalan - `metadata` support and `datasetId` support (NGSI-LD only) is now ready for review.

File: doc/usermanual.md [hunk 1]
```diff
@@ -258,23 +258,25 @@ function setDataUpdateHandler(newHandler)
 ###### Description
 
 Sets the new user handler for Entity update requests. This handler will be called whenever an update request arrives
-with the following parameters: (`id`, `type`, `service`, `subservice`, `attributes`, `callback`). The handler is in charge of
-updating the corresponding values in the devices with the appropriate protocol.
+with the following parameters: (`id`, `type`, `service`, `subservice`, `attributes`, `callback`). Every object within of
+the `attributes` array contains `name`, `type` and `value` attributes, and may also include additional attributes for
+`metadata` and `datasetId` The handler is in charge of updating the corresponding values in the devices with the
```
Comment (fgalan) [inline doc/usermanual.md:263]:
```suggestion
`metadata` and `datasetId`. The handler is in charge of updating the corresponding values in the devices with the
```

File: CHANGES_NEXT_RELEASE [hunk 2]
```diff
@@ -1,2 +1,3 @@
-
-
+- Add datasetId and metadata support for NGSI-LD PATCH+PUT operations (commands)
+- Add datasetId and metadata support for NGSI-LD Notifications (bidirectional attributes)
+- Add metadata support to NGSI-v2 notifications (bidirectional attributes)
```
Comment (fgalan) [inline CHANGES_NEXT_RELEASE:3]:
```suggestion
- Add: datasetId and metadata support for NGSI-LD PATCH+PUT operations (commands)
- Add: datasetId and metadata support for NGSI-LD Notifications (bidirectional attributes)
- Add: metadata support to NGSI-v2 notifications (bidirectional attributes)
```

**Comment by mapedraza** [comment]
LGTM

However, I would like to remark before merging:

- This functionality is outdated. It was not reviewed in deep how compatible is with latest changes in the agent. In fact, expression plugin used in this file is the legacy one:
- https://github.com/telefonicaid/iotagent-node-lib/blob/e7748f20734594df01479ebb356bfdc0e8d38f51/lib/plugins/bidirectionalData.js#L29
- In a future iotagent-node-lib refactor, some functionality that now are plugins will be included as core of the agent (i.e. multientity and expression plugin). This means that, at this point, the viability of the plugin structure after the refactor should be studied in deep. Therefore, there is the possibility that plugins functionality will be dropped if it requires a considerable amount of effort.

**Review by mapedraza — APPROVED** [review]

**Pull request merged**